subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -147,146 +122,172 @@ bt_mesh_dfu_slot_add(size_t size, const uint8_t *fwid,
-		LOG_WRN("Storage failed. err: %d", err);
-		atomic_set_bit_to(valid_slots, idx, prev);
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	return err;
+	slot->slot.fwid_len = fwid_len;
+	memcpy(slot->slot.fwid, fwid, fwid_len);
+	return 0;
 }
 
-bool bt_mesh_dfu_slot_is_valid(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_commit(struct bt_mesh_dfu_slot *dfu_slot)
 {
-	uint16_t idx;
+	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (slot->state != SLOT_STATE_RESERVED ||
+	    slot->slot.fwid_len == 0 ||
+	    slot->slot.size == 0) {
+		return -EINVAL;
+	}
+
+	if (bt_mesh_dfu_slot_get(slot->slot.fwid, slot->slot.fwid_len, NULL)
+	    >= 0) {
+		return -EALREADY;
+	}
+
+	slot->idx = committed_slot_count;
+	slot->state = SLOT_STATE_COMMITTED;
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return false;
+	err = slot_store(slot);
+	if (err) {
+		LOG_WRN("Store failed (err: %d)", err);
+		slot->state = SLOT_STATE_RESERVED;
+		slot->idx = 0;
+		return err;
 	}
 
-	idx = slot_idx(slot);
-	return atomic_test_bit(valid_slots, idx);
+	committed_slot_count++;
+
+	LOG_DBG("Stored slot #%u: %s", slot - &slots[0],
+		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
+	return 0;
 }
 
-int bt_mesh_dfu_slot_del(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_release(const struct bt_mesh_dfu_slot *dfu_slot)
 {
-	struct slot *s = CONTAINER_OF(slot, struct slot, slot);
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+	if (slot->state != SLOT_STATE_RESERVED) {
+		return -EINVAL;
 	}
 
-	LOG_DBG("%u", s - &slots[0]);
-
-	slot_erase(s);
-	slot_invalidate(s);
-	sys_slist_find_and_remove(&list, &s->n);
+	slot->state = SLOT_STATE_FREE;
 
 	return 0;
 }
 
-int bt_mesh_dfu_slot_del_all(void)
+int bt_mesh_dfu_slot_del(const struct bt_mesh_dfu_slot *dfu_slot)
 {
-	struct slot *s;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (slot->state != SLOT_STATE_COMMITTED) {
+		return -EINVAL;
+	}
+
+	int idx = slot - &slots[0];
+
+	LOG_DBG("%u", idx);
 
-	SYS_SLIST_FOR_EACH_CONTAINER(&list, s, n) {
-		slot_erase(s);
-		slot_invalidate(s);
+	slot_erase(slot);
+	slot->state = SLOT_STATE_FREE;
+
+	for (int i = 0; i < CONFIG_BT_MESH_DFU_SLOT_CNT; i++) {
+		if (slots[i].idx > slot->idx) {
+			slots[i].idx--;
+			slot_store(&slots[i]);
+		}
 	}
 
-	sys_slist_init(&list);
+	committed_slot_count--;
 
 	return 0;
 }
 
-const struct bt_mesh_dfu_slot *bt_mesh_dfu_slot_at(uint16_t idx)
+void bt_mesh_dfu_slot_del_all(void)
 {
-	struct slot *s;
+	for (int i = 0; i < CONFIG_BT_MESH_DFU_SLOT_CNT; i++) {
+		if (slots[i].state == SLOT_STATE_COMMITTED) {
+			slot_erase(&slots[i]);
+			slots[i].state = SLOT_STATE_FREE;
+		}
+	}
+
+	committed_slot_count = 0;
+}
 
-	SYS_SLIST_FOR_EACH_CONTAINER(&list, s, n) {
-		if (!idx--) {
-			return &s->slot;
+struct bt_mesh_dfu_slot *bt_mesh_dfu_slot_at(uint16_t idx)
```

Comment:
Well, then why does it doubly checks 
`slots[i].idx == idx` ? Wouldn't it be enough to simply return by using `idx` for array referencing? But that won't work. AFAICS, If the device has four slots, slot_data_load() can assign idx of 3 to slot[0] (depending on where it is positioned in the settings flash since entries can be stored in different order)? This is clear if someone read the relavent code, but it is very easy incorrectly infer that idx actually represents the physical ordering of the slots on the flash by just reading API decriptions and nomenclature. Hence the suggestion.

Reply:
The fact that the slots exist in an array is not supposed to exist outside this file. The only place that a slot has "two indices" (i and idx in this function for example) are inside this file. All APIs use idx to mean slot::idx. The name "index" is also what the spec uses for this field, so it makes sense for outside viewers. I don't feel like changing this. Meaning: Where a slot is in the array is an implementation detail. No one should care and no one outside this file should know. They should not even care that it is an array.

Reply:
And, yes. array location can be different from slot::idx. This is the entire point of having the slot::idx field.

Reply:
My only comment is about the name choosen to represent the `idx`. In all of mesh code base, `idx` variable name is always used to represent some kind of index. Just like firmware has Firmware ID, slot has a Slot ID?

Reply:
No, it's not an ID. An ID I would expect to be constant for the lifetime of the slot. This index is not. It's the "Index" from 4.3.5 in the spec. All places in the API that mention `idx` refer to the "Index" from that chapter. As in: If we order the commited slots in the order they were added, what index of that (abstract) list is the slot at. (This list being the "Firmware Images List" state from 4.3.5). Position in the list may change if an earlier slot is removed. I still can't see why it's confusing to call it index. It's only confusing if you think about the backing array used for storing the slots, but no one should do that, they should just look at the APIs where idx is always == "Index" from 4.3.5.

Reply:
Hmm, now I get it with that. Ok, can this be then `img_idx` or `imglist_idx`?

Reply:
I don't see much point, seeing as there are no other indexes in this API, but sure. I'll change it everywhere then, to be consistent.